{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "742d1eec",
   "metadata": {},
   "source": [
    "**Tempus Coding Challenge:** Annotation Functions for VCF Data\n",
    "\n",
    "*Disclaimer: This documentation is intended to show the thought process for developing this variant interpretation pipeline. For a more concise description of how to use the pipeline, see the README file on GitHub.*\n",
    "\n",
    "Our goal is to take an input VCF and extract some quality control information (such as read depth), along with variant-specific information (such as minor allele frequency, MAF). Examining the VCF file header shows that some of this quality control information is already present in the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "bd48c183",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##INFO=<ID=FR,Number=.,Type=Float,Description=\"Estimated population frequency of variant\">\r\n",
      "##INFO=<ID=MMLQ,Number=1,Type=Float,Description=\"Median minimum base quality for bases around variant\">\r\n",
      "##INFO=<ID=TCR,Number=1,Type=Integer,Description=\"Total reverse strand coverage at this locus\">\r\n",
      "##INFO=<ID=HP,Number=1,Type=Integer,Description=\"Homopolymer run length around variant locus\">\r\n",
      "##INFO=<ID=WE,Number=1,Type=Integer,Description=\"End position of calling window\">\r\n",
      "##INFO=<ID=Source,Number=.,Type=String,Description=\"Was this variant suggested by Playtypus, Assembler, or from a VCF?\">\r\n",
      "##INFO=<ID=FS,Number=.,Type=Float,Description=\"Fisher's exact test for strand bias (Phred scale)\">\r\n",
      "##INFO=<ID=WS,Number=1,Type=Integer,Description=\"Starting position of calling window\">\r\n",
      "##INFO=<ID=PP,Number=.,Type=Float,Description=\"Posterior probability (phred scaled) that this variant segregates\">\r\n",
      "##INFO=<ID=TR,Number=.,Type=Integer,Description=\"Total number of reads containing this variant\">\r\n",
      "##INFO=<ID=NF,Number=.,Type=Integer,Description=\"Total number of forward reads containing this variant\">\r\n",
      "##INFO=<ID=TCF,Number=1,Type=Integer,Description=\"Total forward strand coverage at this locus\">\r\n",
      "##INFO=<ID=NR,Number=.,Type=Integer,Description=\"Total number of reverse reads containing this variant\">\r\n",
      "##INFO=<ID=TC,Number=1,Type=Integer,Description=\"Total coverage at this locus\">\r\n",
      "##INFO=<ID=END,Number=.,Type=Integer,Description=\"End position of reference call block\">\r\n",
      "##INFO=<ID=MGOF,Number=.,Type=Integer,Description=\"Worst goodness-of-fit value reported across all samples\">\r\n",
      "##INFO=<ID=SbPval,Number=.,Type=Float,Description=\"Binomial P-value for strand bias test\">\r\n",
      "##INFO=<ID=START,Number=.,Type=Integer,Description=\"Start position of reference call block\">\r\n",
      "##INFO=<ID=ReadPosRankSum,Number=.,Type=Float,Description=\"Mann-Whitney Rank sum test for difference between in positions of variants in reads from ref and alt\">\r\n",
      "##INFO=<ID=MQ,Number=.,Type=Float,Description=\"Root mean square of mapping qualities of reads at the variant position\">\r\n",
      "##INFO=<ID=QD,Number=1,Type=Float,Description=\"Variant-quality/read-depth for this variant\">\r\n",
      "##INFO=<ID=SC,Number=1,Type=String,Description=\"Genomic sequence 10 bases either side of variant position\">\r\n",
      "##INFO=<ID=BRF,Number=1,Type=Float,Description=\"Fraction of reads around this variant that failed filters\">\r\n",
      "##INFO=<ID=HapScore,Number=.,Type=Integer,Description=\"Haplotype score measuring the number of haplotypes the variant is segregating into in a window\">\r\n",
      "##INFO=<ID=Size,Number=.,Type=Integer,Description=\"Size of reference call block\">\r\n"
     ]
    }
   ],
   "source": [
    "!grep '##INFO' test_vcf_data.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9d8f6a26",
   "metadata": {},
   "source": [
    "As we can see from the header, the INFO column of the VCF contains key information we can reference. We are specifically interested in the following annotations within the INFO column:\n",
    "+ Depth of sequence coverage at a particular site (TC)\n",
    "+ Number of reads supporting a variant (TR)\n",
    "\n",
    "Since VCF files are fundamentally tabular data, we can read them into python dataframes using the Pandas library. Let's write a quick function to do so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d1ae6c6f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1158631</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2965.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.16;FR=1.0000;HP=1;HapScore=1;MGOF=3;MMLQ...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/1:-300.0,-43.88,0.0:3:99:160:156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1246004</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2965.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.09;FR=1.0000;HP=6;HapScore=1;MGOF=5;MMLQ...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/1:-300.0,-41.24,0.0:5:99:152:148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1249187</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2965.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.16;FR=1.0000;HP=3;HapScore=1;MGOF=3;MMLQ...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/1:-300.0,-37.63,0.0:3:99:137:135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1261824</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>2965.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.15;FR=1.0000;HP=1;HapScore=1;MGOF=5;MMLQ...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/1:-300.0,-39.74,0.0:5:99:136:134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1387667</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2965.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.17;FR=1.0000;HP=2;HapScore=1;MGOF=3;MMLQ...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/1:-300.0,-39.41,0.0:3:99:137:133</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS ID REF ALT    QUAL FILTER  \\\n",
       "0     1  1158631  .   A   G  2965.0   PASS   \n",
       "1     1  1246004  .   A   G  2965.0   PASS   \n",
       "2     1  1249187  .   G   A  2965.0   PASS   \n",
       "3     1  1261824  .   G   C  2965.0   PASS   \n",
       "4     1  1387667  .   C   G  2965.0   PASS   \n",
       "\n",
       "                                                INFO              FORMAT  \\\n",
       "0  BRF=0.16;FR=1.0000;HP=1;HapScore=1;MGOF=3;MMLQ...  GT:GL:GOF:GQ:NR:NV   \n",
       "1  BRF=0.09;FR=1.0000;HP=6;HapScore=1;MGOF=5;MMLQ...  GT:GL:GOF:GQ:NR:NV   \n",
       "2  BRF=0.16;FR=1.0000;HP=3;HapScore=1;MGOF=3;MMLQ...  GT:GL:GOF:GQ:NR:NV   \n",
       "3  BRF=0.15;FR=1.0000;HP=1;HapScore=1;MGOF=5;MMLQ...  GT:GL:GOF:GQ:NR:NV   \n",
       "4  BRF=0.17;FR=1.0000;HP=2;HapScore=1;MGOF=3;MMLQ...  GT:GL:GOF:GQ:NR:NV   \n",
       "\n",
       "                               sample  \n",
       "0  1/1:-300.0,-43.88,0.0:3:99:160:156  \n",
       "1  1/1:-300.0,-41.24,0.0:5:99:152:148  \n",
       "2  1/1:-300.0,-37.63,0.0:3:99:137:135  \n",
       "3  1/1:-300.0,-39.74,0.0:5:99:136:134  \n",
       "4  1/1:-300.0,-39.41,0.0:3:99:137:133  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def VCF_to_df(input_path):\n",
    "\n",
    "    # get header\n",
    "    with open(input_path, \"r\") as file:\n",
    "        for line in file.readlines():\n",
    "            if '#CHROM' in line:\n",
    "                header = line.replace('\\n','').replace('#','').split('\\t')\n",
    "                break\n",
    "    \n",
    "    # define dtypes, which helps minimize pandas memory utilization\n",
    "    dtypes = {'CHROM':str,'POS':int,'ID':str,'REF':str,'ALT':str,\n",
    "              'QUAL':float,'FILTER':str,'INFO':str,'FORMAT':str}\n",
    "    \n",
    "    # import overlap file as pandas dataframe\n",
    "    df = pd.read_csv(input_path, sep='\\t', comment='#', names=header, dtype=dtypes)\n",
    "    \n",
    "    return(df)\n",
    "\n",
    "df = VCF_to_df('test_vcf_data.txt')\n",
    "\n",
    "df.head(n=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "448fc7ad",
   "metadata": {},
   "source": [
    "As we can see, the INFO column contains a delimited string of many different annotations. Now that we've imported our dataframe into python, we can extract informaiton depth and supporting reads from INFO to their own columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "16d2e226",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "New columns:\n",
      "['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'sample', 'BRF', 'FR', 'HP', 'HapScore', 'MGOF', 'MMLQ', 'MQ', 'NF', 'NR', 'PP', 'QD', 'SC', 'SbPval', 'Source', 'TC', 'TCF', 'TCR', 'TR', 'WE', 'WS']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depth</th>\n",
       "      <th>var_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>160</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>152</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>137</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>136</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>137</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  depth var_reads\n",
       "0   160       156\n",
       "1   152       148\n",
       "2   137       135\n",
       "3   136       134\n",
       "4   137       133"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explode the delimited INFO column to separate df columns\n",
    "colnames = [i.split('=')[0] for i in df.INFO.iloc[0].split(';')]\n",
    "df[colnames] = df.INFO.apply(lambda row: ';'.join([i.split('=')[1] for i in row.split(';')])).str.split(';',expand=True)\n",
    "\n",
    "# check that we have new dataframe columns\n",
    "print('New columns:')\n",
    "print(df.columns.to_list())\n",
    "\n",
    "# rename our columns of interest to give them more explicit names (rather than acroynms)\n",
    "df.rename(columns={'TC':'depth', 'TR':'var_reads'}, inplace=True)\n",
    "\n",
    "# do a quick visual check of our depth and var_reads columns\n",
    "df[['depth', 'var_reads']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef56daf1",
   "metadata": {},
   "source": [
    "Next, we need to calculate the percentage of reads that support our variant (versus reference). Intuitively, this is simply (var_reads / depth). But if we try to calculate this, we run into our first problem: not all depth and var_reads columns are integers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "48abcbd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>sample</th>\n",
       "      <th>...</th>\n",
       "      <th>QD</th>\n",
       "      <th>SC</th>\n",
       "      <th>SbPval</th>\n",
       "      <th>Source</th>\n",
       "      <th>depth</th>\n",
       "      <th>TCF</th>\n",
       "      <th>TCR</th>\n",
       "      <th>var_reads</th>\n",
       "      <th>WE</th>\n",
       "      <th>WS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1</td>\n",
       "      <td>6219287</td>\n",
       "      <td>.</td>\n",
       "      <td>TCACACA</td>\n",
       "      <td>TCACA,T</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.35;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:6:99:172,172:108,108</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TGAGACTCCATCACACACACA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>172</td>\n",
       "      <td>170</td>\n",
       "      <td>2</td>\n",
       "      <td>108,108</td>\n",
       "      <td>6219302</td>\n",
       "      <td>6219277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>1</td>\n",
       "      <td>91859795</td>\n",
       "      <td>.</td>\n",
       "      <td>TATGTGA</td>\n",
       "      <td>CATGTGA,CATGTGG</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.23;FR=0.5000,0.5000;HP=3;HapScore=2;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:1:99:209,209:208,95</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TTGCCAGCAATATGTGATAAG</td>\n",
       "      <td>0.54</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>209</td>\n",
       "      <td>115</td>\n",
       "      <td>94</td>\n",
       "      <td>208,95</td>\n",
       "      <td>91859809</td>\n",
       "      <td>91859785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>1</td>\n",
       "      <td>154516571</td>\n",
       "      <td>.</td>\n",
       "      <td>TATGCACG</td>\n",
       "      <td>CATGCACA,CATGCACG</td>\n",
       "      <td>2944.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.14;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:199,199:104,191</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TTTTCTGGGCTATGCACGTCA</td>\n",
       "      <td>0.51</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>199</td>\n",
       "      <td>76</td>\n",
       "      <td>123</td>\n",
       "      <td>104,191</td>\n",
       "      <td>154516586</td>\n",
       "      <td>154516561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771</th>\n",
       "      <td>1</td>\n",
       "      <td>156914240</td>\n",
       "      <td>.</td>\n",
       "      <td>CCATT</td>\n",
       "      <td>C,CCATTCATT</td>\n",
       "      <td>2991.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.16;FR=0.5000,0.5000;HP=1;HapScore=3;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:3:99:128,128:66,66</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TAAGGAGACACCATTCATTCA</td>\n",
       "      <td>0.79</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>128</td>\n",
       "      <td>27</td>\n",
       "      <td>101</td>\n",
       "      <td>66,66</td>\n",
       "      <td>156914251</td>\n",
       "      <td>156914230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008</th>\n",
       "      <td>1</td>\n",
       "      <td>182429295</td>\n",
       "      <td>.</td>\n",
       "      <td>CTGTG</td>\n",
       "      <td>C,CTG</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.17;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:201,201:103,103</td>\n",
       "      <td>...</td>\n",
       "      <td>3.36893203883</td>\n",
       "      <td>ATGCGTGTGTCTGTGTGTGTG</td>\n",
       "      <td>0.65</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>201</td>\n",
       "      <td>24</td>\n",
       "      <td>177</td>\n",
       "      <td>103,103</td>\n",
       "      <td>182429306</td>\n",
       "      <td>182429285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1348</th>\n",
       "      <td>1</td>\n",
       "      <td>235543476</td>\n",
       "      <td>.</td>\n",
       "      <td>TTGTGTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>T,TTGTGTGTG</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>alleleBias</td>\n",
       "      <td>BRF=0.09;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:3:99:104,104:17,17</td>\n",
       "      <td>...</td>\n",
       "      <td>142.176470588</td>\n",
       "      <td>TAAGCAATTATTGTGTGTGTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>104</td>\n",
       "      <td>14</td>\n",
       "      <td>90</td>\n",
       "      <td>17,17</td>\n",
       "      <td>235543505</td>\n",
       "      <td>235543466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2626</th>\n",
       "      <td>3</td>\n",
       "      <td>64527465</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>2906.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.24;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:8:99:169,169:75,92</td>\n",
       "      <td>...</td>\n",
       "      <td>26.6636161673</td>\n",
       "      <td>TGAACAAAGACATCCTGTTCT</td>\n",
       "      <td>0.79</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>169</td>\n",
       "      <td>132</td>\n",
       "      <td>37</td>\n",
       "      <td>75,92</td>\n",
       "      <td>64527473</td>\n",
       "      <td>64527455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2779</th>\n",
       "      <td>3</td>\n",
       "      <td>113376110</td>\n",
       "      <td>.</td>\n",
       "      <td>TTGC</td>\n",
       "      <td>T,TTGCTGCTGC</td>\n",
       "      <td>2994.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.17;FR=0.5000,0.5000;HP=1;HapScore=3;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>2/1:-1,-1,-1:5:99:151,151:42,42</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>GTTGTTGCTGTTGCTGCTGCT</td>\n",
       "      <td>0.74</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>151</td>\n",
       "      <td>31</td>\n",
       "      <td>120</td>\n",
       "      <td>42,42</td>\n",
       "      <td>113376121</td>\n",
       "      <td>113376100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3386</th>\n",
       "      <td>4</td>\n",
       "      <td>87024228</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CTGTGTA,CTGTGTGTGTG</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.26;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:8:99:149,141:62,59</td>\n",
       "      <td>...</td>\n",
       "      <td>20.8709677419</td>\n",
       "      <td>GATTTCAACTCTGTGTGTGTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>149</td>\n",
       "      <td>147</td>\n",
       "      <td>2</td>\n",
       "      <td>62,59</td>\n",
       "      <td>87024242</td>\n",
       "      <td>87024218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>4</td>\n",
       "      <td>153884092</td>\n",
       "      <td>.</td>\n",
       "      <td>GGTGTGTGT</td>\n",
       "      <td>G,GGTGTGTGTGT</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.15;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:8:99:113,113:42,42</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TCTATCTAAAGGTGTGTGTGT</td>\n",
       "      <td>0.34</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>113</td>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>42,42</td>\n",
       "      <td>153884109</td>\n",
       "      <td>153884082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3706</th>\n",
       "      <td>4</td>\n",
       "      <td>190394938</td>\n",
       "      <td>.</td>\n",
       "      <td>CAAGAAG</td>\n",
       "      <td>CAAG,C</td>\n",
       "      <td>2991.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.22;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:3:78:140,140:69,69</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>AGGATGGTGACAAGAAGAAGA</td>\n",
       "      <td>0.66</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>140</td>\n",
       "      <td>46</td>\n",
       "      <td>94</td>\n",
       "      <td>69,69</td>\n",
       "      <td>190394951</td>\n",
       "      <td>190394928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4402</th>\n",
       "      <td>6</td>\n",
       "      <td>35477025</td>\n",
       "      <td>.</td>\n",
       "      <td>CTTCTTTA</td>\n",
       "      <td>GTTCTTTA,GTTCTTTG</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.2;FR=0.5000,0.5000;HP=3;HapScore=2;MGOF=...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:3:99:307,307:299,133</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TTTGATTGCTCTTCTTTATCA</td>\n",
       "      <td>0.49</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>307</td>\n",
       "      <td>128</td>\n",
       "      <td>179</td>\n",
       "      <td>299,133</td>\n",
       "      <td>35477040</td>\n",
       "      <td>35477015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4541</th>\n",
       "      <td>6</td>\n",
       "      <td>65016977</td>\n",
       "      <td>.</td>\n",
       "      <td>TGAGA</td>\n",
       "      <td>T,TGA</td>\n",
       "      <td>2994.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.21;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:2:86:184,184:106,106</td>\n",
       "      <td>...</td>\n",
       "      <td>0.77358490566</td>\n",
       "      <td>TACAGGCATCTGAGAGAGAGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>184</td>\n",
       "      <td>15</td>\n",
       "      <td>169</td>\n",
       "      <td>106,106</td>\n",
       "      <td>65016988</td>\n",
       "      <td>65016967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4657</th>\n",
       "      <td>6</td>\n",
       "      <td>109767930</td>\n",
       "      <td>.</td>\n",
       "      <td>GC</td>\n",
       "      <td>TC,TT</td>\n",
       "      <td>2901.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.15;FR=0.5000,0.5000;HP=3;HapScore=2;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:2:99:206,206:203,95</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>GCTGCCTTCCGCTTTGTGGTC</td>\n",
       "      <td>0.48</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>206</td>\n",
       "      <td>82</td>\n",
       "      <td>124</td>\n",
       "      <td>203,95</td>\n",
       "      <td>109767939</td>\n",
       "      <td>109767920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4711</th>\n",
       "      <td>6</td>\n",
       "      <td>123592184</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CTG,CTGTG</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.21;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:6:99:127,127:64,64</td>\n",
       "      <td>...</td>\n",
       "      <td>6.625</td>\n",
       "      <td>CAGTGAAGGGCTGTGTGTGTG</td>\n",
       "      <td>0.99</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>127</td>\n",
       "      <td>118</td>\n",
       "      <td>9</td>\n",
       "      <td>64,64</td>\n",
       "      <td>123592192</td>\n",
       "      <td>123592174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5045</th>\n",
       "      <td>7</td>\n",
       "      <td>57216012</td>\n",
       "      <td>.</td>\n",
       "      <td>ACATTTATT</td>\n",
       "      <td>GAATTTATC,GCATTTATC</td>\n",
       "      <td>2944.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.22;FR=0.5000,0.5000;HP=1;HapScore=2;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:258,258:105,108</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TTCAGTATGTACATTTATTTC</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>258</td>\n",
       "      <td>56</td>\n",
       "      <td>202</td>\n",
       "      <td>105,108</td>\n",
       "      <td>57216028</td>\n",
       "      <td>57216002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5847</th>\n",
       "      <td>9</td>\n",
       "      <td>4850421</td>\n",
       "      <td>.</td>\n",
       "      <td>CCCTTGGGCCA</td>\n",
       "      <td>GCCTTGGGCCA,GCCTTGGGCCG</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.16;FR=0.5000,0.5000;HP=2;HapScore=4;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:8:99:131,131:120,49</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>GGGCTATGCACCCTTGGGCCA</td>\n",
       "      <td>0.51</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>131</td>\n",
       "      <td>3</td>\n",
       "      <td>128</td>\n",
       "      <td>120,49</td>\n",
       "      <td>4850444</td>\n",
       "      <td>4850411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874</th>\n",
       "      <td>9</td>\n",
       "      <td>14775853</td>\n",
       "      <td>.</td>\n",
       "      <td>ATCTGTG</td>\n",
       "      <td>GTCTGTA,GTCTGTG</td>\n",
       "      <td>2944.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.12;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:175,175:86,170</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>GGTTTGTCCCATCTGTGGCCA</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>175</td>\n",
       "      <td>109</td>\n",
       "      <td>66</td>\n",
       "      <td>86,170</td>\n",
       "      <td>14775867</td>\n",
       "      <td>14775843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>10</td>\n",
       "      <td>47087608</td>\n",
       "      <td>.</td>\n",
       "      <td>TG</td>\n",
       "      <td>CA,CG</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.1;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF=...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:12:99:149,149:38,147</td>\n",
       "      <td>...</td>\n",
       "      <td>25.8017861806</td>\n",
       "      <td>TGGCCTTTGCTGTGCTCTGGC</td>\n",
       "      <td>0.53</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>149</td>\n",
       "      <td>122</td>\n",
       "      <td>27</td>\n",
       "      <td>38,147</td>\n",
       "      <td>47087617</td>\n",
       "      <td>47087598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7151</th>\n",
       "      <td>11</td>\n",
       "      <td>63673603</td>\n",
       "      <td>.</td>\n",
       "      <td>CTGTG</td>\n",
       "      <td>C,CTG</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.22;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:3:99:267,267:148,148</td>\n",
       "      <td>...</td>\n",
       "      <td>4.00675675676</td>\n",
       "      <td>GTTGAGCCCGCTGTGTGTGTG</td>\n",
       "      <td>0.61</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>267</td>\n",
       "      <td>129</td>\n",
       "      <td>138</td>\n",
       "      <td>148,148</td>\n",
       "      <td>63673614</td>\n",
       "      <td>63673593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7248</th>\n",
       "      <td>11</td>\n",
       "      <td>85979636</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTGT</td>\n",
       "      <td>C,CGT</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.19;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:115,115:56,56</td>\n",
       "      <td>...</td>\n",
       "      <td>9.07142857143</td>\n",
       "      <td>GTACTTCCATCGTGTGTGTGT</td>\n",
       "      <td>0.99</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>115</td>\n",
       "      <td>25</td>\n",
       "      <td>90</td>\n",
       "      <td>56,56</td>\n",
       "      <td>85979647</td>\n",
       "      <td>85979626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8077</th>\n",
       "      <td>13</td>\n",
       "      <td>46108853</td>\n",
       "      <td>.</td>\n",
       "      <td>TG</td>\n",
       "      <td>CA,CG</td>\n",
       "      <td>2944.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.16;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:3:99:188,188:109,188</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>CTTCTGCTGTTGGTTTCTAAA</td>\n",
       "      <td>0.54</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>188</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>109,188</td>\n",
       "      <td>46108862</td>\n",
       "      <td>46108843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8199</th>\n",
       "      <td>13</td>\n",
       "      <td>103705044</td>\n",
       "      <td>.</td>\n",
       "      <td>AAACCAG</td>\n",
       "      <td>CAACCAA,CAACCAG</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.22;FR=0.5000,0.5000;HP=2;HapScore=2;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:3:99:233,233:98,232</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>ACAACGAGAGAAACCAGAGAT</td>\n",
       "      <td>0.93</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>233</td>\n",
       "      <td>61</td>\n",
       "      <td>172</td>\n",
       "      <td>98,232</td>\n",
       "      <td>103705058</td>\n",
       "      <td>103705034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9162</th>\n",
       "      <td>16</td>\n",
       "      <td>11214668</td>\n",
       "      <td>.</td>\n",
       "      <td>AAC</td>\n",
       "      <td>AACAC,A</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.19;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>2/1:-1,-1,-1:10:99:178,178:103,103</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>GTTGGCTACAAACACACACAC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>178</td>\n",
       "      <td>17</td>\n",
       "      <td>161</td>\n",
       "      <td>103,103</td>\n",
       "      <td>11214679</td>\n",
       "      <td>11214645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9351</th>\n",
       "      <td>16</td>\n",
       "      <td>88585145</td>\n",
       "      <td>.</td>\n",
       "      <td>GTGGATGGATGGA</td>\n",
       "      <td>G,GTGGATGGA</td>\n",
       "      <td>2992.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.47;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:292,292:88,88</td>\n",
       "      <td>...</td>\n",
       "      <td>18.6136363636</td>\n",
       "      <td>GGCTGAGAGGGTGGATGGATG</td>\n",
       "      <td>0.99</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>292</td>\n",
       "      <td>187</td>\n",
       "      <td>105</td>\n",
       "      <td>88,88</td>\n",
       "      <td>88585186</td>\n",
       "      <td>88585135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9574</th>\n",
       "      <td>17</td>\n",
       "      <td>30791503</td>\n",
       "      <td>.</td>\n",
       "      <td>CCCCCCC</td>\n",
       "      <td>TCCCCCC,TCCCCCG</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.26;FR=0.5000,0.5000;HP=7;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:2:99:324,324:317,78</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>GTGGCTTTTTCCCCCCCCTTC</td>\n",
       "      <td>0.58</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>324</td>\n",
       "      <td>214</td>\n",
       "      <td>110</td>\n",
       "      <td>317,78</td>\n",
       "      <td>30791517</td>\n",
       "      <td>30791492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9669</th>\n",
       "      <td>17</td>\n",
       "      <td>39406409</td>\n",
       "      <td>.</td>\n",
       "      <td>CCAGCTGCTGTCAGCCTTT</td>\n",
       "      <td>ACAGCTGCTGCCAGCCTTC,ACAGCTGCTGTCAGCCTTT</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.25;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:27:99:370,370:41,365</td>\n",
       "      <td>...</td>\n",
       "      <td>62.0845823137</td>\n",
       "      <td>ACCTGTGTGTCCAGCTGCTGT</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>370</td>\n",
       "      <td>165</td>\n",
       "      <td>205</td>\n",
       "      <td>41,365</td>\n",
       "      <td>39406435</td>\n",
       "      <td>39406399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9676</th>\n",
       "      <td>17</td>\n",
       "      <td>39594934</td>\n",
       "      <td>.</td>\n",
       "      <td>AC</td>\n",
       "      <td>A,ACC</td>\n",
       "      <td>2913.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.16;FR=0.5000,0.5000;HP=8;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:195,195:151,151</td>\n",
       "      <td>...</td>\n",
       "      <td>6.54304635762</td>\n",
       "      <td>CTGGTTCTATACCCCCCCCAC</td>\n",
       "      <td>0.4</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>195</td>\n",
       "      <td>128</td>\n",
       "      <td>67</td>\n",
       "      <td>151,151</td>\n",
       "      <td>39594942</td>\n",
       "      <td>39594923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9815</th>\n",
       "      <td>17</td>\n",
       "      <td>53124546</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTGTGT</td>\n",
       "      <td>C,CGTGTGTGT</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.18;FR=0.5000,0.5000;HP=1;HapScore=2;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:5:99:140,140:50,50</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>TATATGTGCTCGTGTGTGTGT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>140</td>\n",
       "      <td>33</td>\n",
       "      <td>107</td>\n",
       "      <td>50,50</td>\n",
       "      <td>53124561</td>\n",
       "      <td>53124533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10118</th>\n",
       "      <td>18</td>\n",
       "      <td>19119974</td>\n",
       "      <td>.</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G,GAT</td>\n",
       "      <td>2991.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.24;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:2:99:140,140:87,87</td>\n",
       "      <td>...</td>\n",
       "      <td>3.88505747126</td>\n",
       "      <td>TCCAGCCCTAGATATATATAT</td>\n",
       "      <td>0.74</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>140</td>\n",
       "      <td>16</td>\n",
       "      <td>124</td>\n",
       "      <td>87,87</td>\n",
       "      <td>19119985</td>\n",
       "      <td>19119964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10191</th>\n",
       "      <td>18</td>\n",
       "      <td>52585071</td>\n",
       "      <td>.</td>\n",
       "      <td>GACACACACAC</td>\n",
       "      <td>GAC,G</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.26;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:179,179:58,58</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>CACGTGCACAGACACACACAC</td>\n",
       "      <td>0.9</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>179</td>\n",
       "      <td>125</td>\n",
       "      <td>54</td>\n",
       "      <td>58,58</td>\n",
       "      <td>52585108</td>\n",
       "      <td>52585061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11436</th>\n",
       "      <td>X</td>\n",
       "      <td>1746526</td>\n",
       "      <td>.</td>\n",
       "      <td>CGAAATGAAATGAAAT</td>\n",
       "      <td>C,TGAAATGAAATGAAAT</td>\n",
       "      <td>2989.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.22;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:7:99:112,89:13,31</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>CGAAATGAAACGAAATGAAAT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>112</td>\n",
       "      <td>102</td>\n",
       "      <td>10</td>\n",
       "      <td>13,31</td>\n",
       "      <td>1746549</td>\n",
       "      <td>1746516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11466</th>\n",
       "      <td>X</td>\n",
       "      <td>12848005</td>\n",
       "      <td>.</td>\n",
       "      <td>TTTCA</td>\n",
       "      <td>T,TTTCATTCA</td>\n",
       "      <td>2991.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.11;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:5:99:110,110:59,59</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>GTTGGTCTTGTTTCATTCATT</td>\n",
       "      <td>0.43</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>110</td>\n",
       "      <td>42</td>\n",
       "      <td>68</td>\n",
       "      <td>59,59</td>\n",
       "      <td>12848016</td>\n",
       "      <td>12847995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11570</th>\n",
       "      <td>X</td>\n",
       "      <td>52895579</td>\n",
       "      <td>.</td>\n",
       "      <td>GGTGT</td>\n",
       "      <td>G,GGT</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.16;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:99:150,150:73,73</td>\n",
       "      <td>...</td>\n",
       "      <td>8.47945205479</td>\n",
       "      <td>TGGAACCAGGGGTGTGTGTGT</td>\n",
       "      <td>0.98</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>150</td>\n",
       "      <td>17</td>\n",
       "      <td>133</td>\n",
       "      <td>73,73</td>\n",
       "      <td>52895592</td>\n",
       "      <td>52895569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11584</th>\n",
       "      <td>X</td>\n",
       "      <td>56295958</td>\n",
       "      <td>.</td>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG</td>\n",
       "      <td>2994.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.22;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:5:99:131,131:75,75</td>\n",
       "      <td>...</td>\n",
       "      <td>2.70666666667</td>\n",
       "      <td>GTCTTTGTGTATGTGTGTGTG</td>\n",
       "      <td>0.84</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>131</td>\n",
       "      <td>22</td>\n",
       "      <td>109</td>\n",
       "      <td>75,75</td>\n",
       "      <td>56295969</td>\n",
       "      <td>56295948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11690</th>\n",
       "      <td>X</td>\n",
       "      <td>122575672</td>\n",
       "      <td>.</td>\n",
       "      <td>AACAC</td>\n",
       "      <td>A,AAC</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRF=0.15;FR=0.4892,0.5108;HP=1;HapScore=1;MGOF...</td>\n",
       "      <td>GT:GL:GOF:GQ:NR:NV</td>\n",
       "      <td>1/2:-1,-1,-1:4:17:150,150:91,91</td>\n",
       "      <td>...</td>\n",
       "      <td>0.153846153846</td>\n",
       "      <td>CCTTCCTCTGAACACACACAC</td>\n",
       "      <td>0.82</td>\n",
       "      <td>Platypus</td>\n",
       "      <td>150</td>\n",
       "      <td>34</td>\n",
       "      <td>116</td>\n",
       "      <td>91,91</td>\n",
       "      <td>122575685</td>\n",
       "      <td>122575662</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS ID                      REF  \\\n",
       "39        1    6219287  .                  TCACACA   \n",
       "414       1   91859795  .                  TATGTGA   \n",
       "713       1  154516571  .                 TATGCACG   \n",
       "771       1  156914240  .                    CCATT   \n",
       "1008      1  182429295  .                    CTGTG   \n",
       "1348      1  235543476  .  TTGTGTGTGTGTGTGTGTGTGTG   \n",
       "2626      3   64527465  .                        C   \n",
       "2779      3  113376110  .                     TTGC   \n",
       "3386      4   87024228  .                        C   \n",
       "3577      4  153884092  .                GGTGTGTGT   \n",
       "3706      4  190394938  .                  CAAGAAG   \n",
       "4402      6   35477025  .                 CTTCTTTA   \n",
       "4541      6   65016977  .                    TGAGA   \n",
       "4657      6  109767930  .                       GC   \n",
       "4711      6  123592184  .                        C   \n",
       "5045      7   57216012  .                ACATTTATT   \n",
       "5847      9    4850421  .              CCCTTGGGCCA   \n",
       "5874      9   14775853  .                  ATCTGTG   \n",
       "6344     10   47087608  .                       TG   \n",
       "7151     11   63673603  .                    CTGTG   \n",
       "7248     11   85979636  .                    CGTGT   \n",
       "8077     13   46108853  .                       TG   \n",
       "8199     13  103705044  .                  AAACCAG   \n",
       "9162     16   11214668  .                      AAC   \n",
       "9351     16   88585145  .            GTGGATGGATGGA   \n",
       "9574     17   30791503  .                  CCCCCCC   \n",
       "9669     17   39406409  .      CCAGCTGCTGTCAGCCTTT   \n",
       "9676     17   39594934  .                       AC   \n",
       "9815     17   53124546  .                  CGTGTGT   \n",
       "10118    18   19119974  .                    GATAT   \n",
       "10191    18   52585071  .              GACACACACAC   \n",
       "11436     X    1746526  .         CGAAATGAAATGAAAT   \n",
       "11466     X   12848005  .                    TTTCA   \n",
       "11570     X   52895579  .                    GGTGT   \n",
       "11584     X   56295958  .                    ATGTG   \n",
       "11690     X  122575672  .                    AACAC   \n",
       "\n",
       "                                           ALT    QUAL      FILTER  \\\n",
       "39                                     TCACA,T  2997.0        PASS   \n",
       "414                            CATGTGA,CATGTGG  2962.0        PASS   \n",
       "713                          CATGCACA,CATGCACG  2944.0        PASS   \n",
       "771                                C,CCATTCATT  2991.0        PASS   \n",
       "1008                                     C,CTG  2995.0        PASS   \n",
       "1348                               T,TTGTGTGTG  2997.0  alleleBias   \n",
       "2626                                       A,T  2906.0        PASS   \n",
       "2779                              T,TTGCTGCTGC  2994.0        PASS   \n",
       "3386                       CTGTGTA,CTGTGTGTGTG  2997.0        PASS   \n",
       "3577                             G,GGTGTGTGTGT  2997.0        PASS   \n",
       "3706                                    CAAG,C  2991.0        PASS   \n",
       "4402                         GTTCTTTA,GTTCTTTG  2962.0        PASS   \n",
       "4541                                     T,TGA  2994.0        PASS   \n",
       "4657                                     TC,TT  2901.0        PASS   \n",
       "4711                                 CTG,CTGTG  2995.0        PASS   \n",
       "5045                       GAATTTATC,GCATTTATC  2944.0        PASS   \n",
       "5847                   GCCTTGGGCCA,GCCTTGGGCCG  2962.0        PASS   \n",
       "5874                           GTCTGTA,GTCTGTG  2944.0        PASS   \n",
       "6344                                     CA,CG  2962.0        PASS   \n",
       "7151                                     C,CTG  2995.0        PASS   \n",
       "7248                                     C,CGT  2997.0        PASS   \n",
       "8077                                     CA,CG  2944.0        PASS   \n",
       "8199                           CAACCAA,CAACCAG  2962.0        PASS   \n",
       "9162                                   AACAC,A  2997.0        PASS   \n",
       "9351                               G,GTGGATGGA  2992.0        PASS   \n",
       "9574                           TCCCCCC,TCCCCCG  2962.0        PASS   \n",
       "9669   ACAGCTGCTGCCAGCCTTC,ACAGCTGCTGTCAGCCTTT  2962.0        PASS   \n",
       "9676                                     A,ACC  2913.0        PASS   \n",
       "9815                               C,CGTGTGTGT  2997.0        PASS   \n",
       "10118                                    G,GAT  2991.0        PASS   \n",
       "10191                                    GAC,G  2997.0        PASS   \n",
       "11436                       C,TGAAATGAAATGAAAT  2989.0        PASS   \n",
       "11466                              T,TTTCATTCA  2991.0        PASS   \n",
       "11570                                    G,GGT  2997.0        PASS   \n",
       "11584                                    A,ATG  2994.0        PASS   \n",
       "11690                                    A,AAC  2995.0        PASS   \n",
       "\n",
       "                                                    INFO              FORMAT  \\\n",
       "39     BRF=0.35;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "414    BRF=0.23;FR=0.5000,0.5000;HP=3;HapScore=2;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "713    BRF=0.14;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "771    BRF=0.16;FR=0.5000,0.5000;HP=1;HapScore=3;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "1008   BRF=0.17;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "1348   BRF=0.09;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "2626   BRF=0.24;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "2779   BRF=0.17;FR=0.5000,0.5000;HP=1;HapScore=3;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "3386   BRF=0.26;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "3577   BRF=0.15;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "3706   BRF=0.22;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "4402   BRF=0.2;FR=0.5000,0.5000;HP=3;HapScore=2;MGOF=...  GT:GL:GOF:GQ:NR:NV   \n",
       "4541   BRF=0.21;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "4657   BRF=0.15;FR=0.5000,0.5000;HP=3;HapScore=2;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "4711   BRF=0.21;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "5045   BRF=0.22;FR=0.5000,0.5000;HP=1;HapScore=2;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "5847   BRF=0.16;FR=0.5000,0.5000;HP=2;HapScore=4;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "5874   BRF=0.12;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "6344   BRF=0.1;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF=...  GT:GL:GOF:GQ:NR:NV   \n",
       "7151   BRF=0.22;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "7248   BRF=0.19;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "8077   BRF=0.16;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "8199   BRF=0.22;FR=0.5000,0.5000;HP=2;HapScore=2;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "9162   BRF=0.19;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "9351   BRF=0.47;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "9574   BRF=0.26;FR=0.5000,0.5000;HP=7;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "9669   BRF=0.25;FR=0.5000,0.5000;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "9676   BRF=0.16;FR=0.5000,0.5000;HP=8;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "9815   BRF=0.18;FR=0.5000,0.5000;HP=1;HapScore=2;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "10118  BRF=0.24;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "10191  BRF=0.26;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "11436  BRF=0.22;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "11466  BRF=0.11;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "11570  BRF=0.16;FR=0.5000,0.5000;HP=3;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "11584  BRF=0.22;FR=0.5000,0.5000;HP=2;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "11690  BRF=0.15;FR=0.4892,0.5108;HP=1;HapScore=1;MGOF...  GT:GL:GOF:GQ:NR:NV   \n",
       "\n",
       "                                   sample  ...              QD  \\\n",
       "39      1/2:-1,-1,-1:6:99:172,172:108,108  ...              20   \n",
       "414      1/2:-1,-1,-1:1:99:209,209:208,95  ...              20   \n",
       "713     1/2:-1,-1,-1:4:99:199,199:104,191  ...              20   \n",
       "771       1/2:-1,-1,-1:3:99:128,128:66,66  ...              20   \n",
       "1008    1/2:-1,-1,-1:4:99:201,201:103,103  ...   3.36893203883   \n",
       "1348      1/2:-1,-1,-1:3:99:104,104:17,17  ...   142.176470588   \n",
       "2626      1/2:-1,-1,-1:8:99:169,169:75,92  ...   26.6636161673   \n",
       "2779      2/1:-1,-1,-1:5:99:151,151:42,42  ...              20   \n",
       "3386      1/2:-1,-1,-1:8:99:149,141:62,59  ...   20.8709677419   \n",
       "3577      1/2:-1,-1,-1:8:99:113,113:42,42  ...              20   \n",
       "3706      1/2:-1,-1,-1:3:78:140,140:69,69  ...              20   \n",
       "4402    1/2:-1,-1,-1:3:99:307,307:299,133  ...              20   \n",
       "4541    1/2:-1,-1,-1:2:86:184,184:106,106  ...   0.77358490566   \n",
       "4657     1/2:-1,-1,-1:2:99:206,206:203,95  ...              20   \n",
       "4711      1/2:-1,-1,-1:6:99:127,127:64,64  ...           6.625   \n",
       "5045    1/2:-1,-1,-1:4:99:258,258:105,108  ...              20   \n",
       "5847     1/2:-1,-1,-1:8:99:131,131:120,49  ...              20   \n",
       "5874     1/2:-1,-1,-1:4:99:175,175:86,170  ...              20   \n",
       "6344    1/2:-1,-1,-1:12:99:149,149:38,147  ...   25.8017861806   \n",
       "7151    1/2:-1,-1,-1:3:99:267,267:148,148  ...   4.00675675676   \n",
       "7248      1/2:-1,-1,-1:4:99:115,115:56,56  ...   9.07142857143   \n",
       "8077    1/2:-1,-1,-1:3:99:188,188:109,188  ...              20   \n",
       "8199     1/2:-1,-1,-1:3:99:233,233:98,232  ...              20   \n",
       "9162   2/1:-1,-1,-1:10:99:178,178:103,103  ...              20   \n",
       "9351      1/2:-1,-1,-1:4:99:292,292:88,88  ...   18.6136363636   \n",
       "9574     1/2:-1,-1,-1:2:99:324,324:317,78  ...              20   \n",
       "9669    1/2:-1,-1,-1:27:99:370,370:41,365  ...   62.0845823137   \n",
       "9676    1/2:-1,-1,-1:4:99:195,195:151,151  ...   6.54304635762   \n",
       "9815      1/2:-1,-1,-1:5:99:140,140:50,50  ...              20   \n",
       "10118     1/2:-1,-1,-1:2:99:140,140:87,87  ...   3.88505747126   \n",
       "10191     1/2:-1,-1,-1:4:99:179,179:58,58  ...              20   \n",
       "11436      1/2:-1,-1,-1:7:99:112,89:13,31  ...              20   \n",
       "11466     1/2:-1,-1,-1:5:99:110,110:59,59  ...              20   \n",
       "11570     1/2:-1,-1,-1:4:99:150,150:73,73  ...   8.47945205479   \n",
       "11584     1/2:-1,-1,-1:5:99:131,131:75,75  ...   2.70666666667   \n",
       "11690     1/2:-1,-1,-1:4:17:150,150:91,91  ...  0.153846153846   \n",
       "\n",
       "                          SC SbPval    Source depth  TCF  TCR var_reads  \\\n",
       "39     TGAGACTCCATCACACACACA    1.0  Platypus   172  170    2   108,108   \n",
       "414    TTGCCAGCAATATGTGATAAG   0.54  Platypus   209  115   94    208,95   \n",
       "713    TTTTCTGGGCTATGCACGTCA   0.51  Platypus   199   76  123   104,191   \n",
       "771    TAAGGAGACACCATTCATTCA   0.79  Platypus   128   27  101     66,66   \n",
       "1008   ATGCGTGTGTCTGTGTGTGTG   0.65  Platypus   201   24  177   103,103   \n",
       "1348   TAAGCAATTATTGTGTGTGTG    1.0  Platypus   104   14   90     17,17   \n",
       "2626   TGAACAAAGACATCCTGTTCT   0.79  Platypus   169  132   37     75,92   \n",
       "2779   GTTGTTGCTGTTGCTGCTGCT   0.74  Platypus   151   31  120     42,42   \n",
       "3386   GATTTCAACTCTGTGTGTGTG    1.0  Platypus   149  147    2     62,59   \n",
       "3577   TCTATCTAAAGGTGTGTGTGT   0.34  Platypus   113   50   63     42,42   \n",
       "3706   AGGATGGTGACAAGAAGAAGA   0.66  Platypus   140   46   94     69,69   \n",
       "4402   TTTGATTGCTCTTCTTTATCA   0.49  Platypus   307  128  179   299,133   \n",
       "4541   TACAGGCATCTGAGAGAGAGA    1.0  Platypus   184   15  169   106,106   \n",
       "4657   GCTGCCTTCCGCTTTGTGGTC   0.48  Platypus   206   82  124    203,95   \n",
       "4711   CAGTGAAGGGCTGTGTGTGTG   0.99  Platypus   127  118    9     64,64   \n",
       "5045   TTCAGTATGTACATTTATTTC   0.75  Platypus   258   56  202   105,108   \n",
       "5847   GGGCTATGCACCCTTGGGCCA   0.51  Platypus   131    3  128    120,49   \n",
       "5874   GGTTTGTCCCATCTGTGGCCA   0.75  Platypus   175  109   66    86,170   \n",
       "6344   TGGCCTTTGCTGTGCTCTGGC   0.53  Platypus   149  122   27    38,147   \n",
       "7151   GTTGAGCCCGCTGTGTGTGTG   0.61  Platypus   267  129  138   148,148   \n",
       "7248   GTACTTCCATCGTGTGTGTGT   0.99  Platypus   115   25   90     56,56   \n",
       "8077   CTTCTGCTGTTGGTTTCTAAA   0.54  Platypus   188   94   94   109,188   \n",
       "8199   ACAACGAGAGAAACCAGAGAT   0.93  Platypus   233   61  172    98,232   \n",
       "9162   GTTGGCTACAAACACACACAC    1.0  Platypus   178   17  161   103,103   \n",
       "9351   GGCTGAGAGGGTGGATGGATG   0.99  Platypus   292  187  105     88,88   \n",
       "9574   GTGGCTTTTTCCCCCCCCTTC   0.58  Platypus   324  214  110    317,78   \n",
       "9669   ACCTGTGTGTCCAGCTGCTGT   0.01  Platypus   370  165  205    41,365   \n",
       "9676   CTGGTTCTATACCCCCCCCAC    0.4  Platypus   195  128   67   151,151   \n",
       "9815   TATATGTGCTCGTGTGTGTGT    1.0  Platypus   140   33  107     50,50   \n",
       "10118  TCCAGCCCTAGATATATATAT   0.74  Platypus   140   16  124     87,87   \n",
       "10191  CACGTGCACAGACACACACAC    0.9  Platypus   179  125   54     58,58   \n",
       "11436  CGAAATGAAACGAAATGAAAT    1.0  Platypus   112  102   10     13,31   \n",
       "11466  GTTGGTCTTGTTTCATTCATT   0.43  Platypus   110   42   68     59,59   \n",
       "11570  TGGAACCAGGGGTGTGTGTGT   0.98  Platypus   150   17  133     73,73   \n",
       "11584  GTCTTTGTGTATGTGTGTGTG   0.84  Platypus   131   22  109     75,75   \n",
       "11690  CCTTCCTCTGAACACACACAC   0.82  Platypus   150   34  116     91,91   \n",
       "\n",
       "              WE         WS  \n",
       "39       6219302    6219277  \n",
       "414     91859809   91859785  \n",
       "713    154516586  154516561  \n",
       "771    156914251  156914230  \n",
       "1008   182429306  182429285  \n",
       "1348   235543505  235543466  \n",
       "2626    64527473   64527455  \n",
       "2779   113376121  113376100  \n",
       "3386    87024242   87024218  \n",
       "3577   153884109  153884082  \n",
       "3706   190394951  190394928  \n",
       "4402    35477040   35477015  \n",
       "4541    65016988   65016967  \n",
       "4657   109767939  109767920  \n",
       "4711   123592192  123592174  \n",
       "5045    57216028   57216002  \n",
       "5847     4850444    4850411  \n",
       "5874    14775867   14775843  \n",
       "6344    47087617   47087598  \n",
       "7151    63673614   63673593  \n",
       "7248    85979647   85979626  \n",
       "8077    46108862   46108843  \n",
       "8199   103705058  103705034  \n",
       "9162    11214679   11214645  \n",
       "9351    88585186   88585135  \n",
       "9574    30791517   30791492  \n",
       "9669    39406435   39406399  \n",
       "9676    39594942   39594923  \n",
       "9815    53124561   53124533  \n",
       "10118   19119985   19119964  \n",
       "10191   52585108   52585061  \n",
       "11436    1746549    1746516  \n",
       "11466   12848016   12847995  \n",
       "11570   52895592   52895569  \n",
       "11584   56295969   56295948  \n",
       "11690  122575685  122575662  \n",
       "\n",
       "[36 rows x 30 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.var_reads.str.contains(',')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08f17f0d",
   "metadata": {},
   "source": [
    "What we see above are sites where multiple ALT alleles have been mapped to a single locus. This is a quirk of some VCF files, but since we are interested in building a *variant* annotator rather than a tool that provides purely positional information, we need to find a way to handle this.\n",
    "\n",
    "We can build a function to handle this for us. Essentially, it will take our dataframe, identify multiallelic variants, and separate them into their own rows. It will not only separate out the two ALT alleles, but also do this for other comma-delimited information like depth and var_reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7536aaf4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expanding multiallelic variants added 36 rows to our data.\n"
     ]
    }
   ],
   "source": [
    "def expand_multiallelic(df):\n",
    "    \n",
    "    # takes a pandas dataframe with VCF 4.0 columns\n",
    "    # explodes multiallelic annotations to new dataframe rows\n",
    "\n",
    "    # create empty dataframe to hold exploded VCF info\n",
    "    expanded = pd.DataFrame(columns=df.columns)\n",
    "\n",
    "    for column in ['REF','ALT']:\n",
    "\n",
    "        # iterate through multiallelic annotations of fixed size (n=2,3,... annotations in a column)\n",
    "        for length in df[column].str.split(',').str.len().unique():\n",
    "\n",
    "            if length > 1:\n",
    "\n",
    "                # subset data\n",
    "                multi = df.loc[df[column].str.contains(',')]\n",
    "                multi = multi.loc[multi[column].str.split(',').str.len()==length]\n",
    "\n",
    "                # duplicate index N times\n",
    "                prevdex = multi.index.repeat(length)\n",
    "\n",
    "                def boolcheck(array):\n",
    "                    # check if a series contains only 1 unique value\n",
    "                    # return the single unique value if so\n",
    "                    if array.size==1:\n",
    "                        return(array[0])\n",
    "                    else:\n",
    "                        return(False)\n",
    "\n",
    "                # select all columns with incorrect number of delimiters - do not explode these\n",
    "                icols = [i for i in multi.columns if boolcheck(multi[i].astype(str).str.count(',').unique())!=length-1 ]\n",
    "                icols = icols + [i for i in ['FORMAT', 'INFO'] if i not in icols]\n",
    "                dcols = [i for i in df.columns if i not in icols]\n",
    "\n",
    "                # set non-delimited columns as index\n",
    "                multi.set_index(icols, inplace=True)\n",
    "\n",
    "                # split all remaining columns apart\n",
    "                multi = pd.DataFrame( [multi[col].str.split(',') for col in multi.columns] ).transpose()\n",
    "                multi = multi.apply(pd.Series.explode).reset_index()\n",
    "                multi.set_index(prevdex, inplace=True)\n",
    "\n",
    "                # concat exploded columns to expanded dataframe\n",
    "                expanded = pd.concat([expanded, multi])\n",
    "\n",
    "    # merge all exploded variants back to original dataframe, and sort\n",
    "    df = pd.concat( [ df.loc[~( df.REF.str.contains(',') | df.ALT.str.contains(','))], expanded ])\n",
    "    df.sort_index(inplace=True)\n",
    "\n",
    "    return(df)\n",
    "\n",
    "# initial df size\n",
    "size1 = df.shape[0]\n",
    "\n",
    "# now use our function to expand multiallelic calls\n",
    "df = expand_multiallelic(df)\n",
    "\n",
    "sizechange = (df.shape[0]) - size1\n",
    "print('Expanding multiallelic variants added', sizechange, 'rows to our data.')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d6dfa1",
   "metadata": {},
   "source": [
    "Now that we've handled our multiallelic sites, we can calculate the percent of overall reads at a site (depth) that contain our variant (var_reads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "6bc0f4da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.975000\n",
       "1    0.973684\n",
       "2    0.985401\n",
       "3    0.985294\n",
       "4    0.970803\n",
       "Name: var_perc, dtype: float64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['var_perc'] = df.var_reads.astype(int) / df.depth.astype(int)\n",
    "df.var_perc.head(n=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "431cdb43",
   "metadata": {},
   "source": [
    "Before we move on to annotating this file with variant-specific information, we can do a quick check to make sure our analysis has not produced missing values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "c174cd37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM        0\n",
       "POS          0\n",
       "REF          0\n",
       "ALT          0\n",
       "depth        0\n",
       "var_reads    0\n",
       "var_perc     0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['CHROM','POS','REF','ALT','depth','var_reads','var_perc']].isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91c504cf",
   "metadata": {},
   "source": [
    "Looks good! Now we can move on to adding variant-specific information. \n",
    "\n",
    "**Variant Specific Annotations**\n",
    "\n",
    "Now that we have preliminary information on variant depth, we can move onto annotating this file with the Ensembl Variant Effect Predictor (VEP). We can use the VEP API to map our variants to affected genes and descriptions of their molecular consequences. However, querying this resource can be slow, as our API requests are limited to 200 variants at a time. Ordinarily, this is work I prefer to do using a local VEP install, but there are certain advantages to API requests. For example, if we have a small VCF file, we can retrieve VEP annotations relatively quickly and without having to perform a time-consuming installation of VEP. Our script is also much more portable to users who may not have VEP installed.\n",
    "\n",
    "We want to retrieve the following VEP annotations:\n",
    "+ Gene(s) affected by the variant (gene_id and gene_symbol)\n",
    "+ Type of variant (SNP, indel, CNV, etc, which is 'variant_class' in the VEP API)\n",
    "+ Molecular consequence of variant (missense, intergenic, 3'UTR, etc)\n",
    "\n",
    "For this last bulletpoint, we have several options we can consider. We could settle for a simple description of possible consequences like \"missense\" or \"splicing.\" However, a single variant can affect multiple overlapping transcripts in different ways. The visualization below, of alternate splicing in the KRAS gene (adapted from the GTExPortal consortium) depicts this nicely. It's very clear that a variant in one exon may be a noncoding variant in another.\n",
    "![KRAS splicing from the GTExPortal](KRAS_transcripts.jpg)\n",
    "\n",
    "For the sake of specificity, and presenting information most relevant for clinical diagnostics or prognostics, I am opting to present the full HGVS variant consequences. This information is present in VEP as \n",
    "`transcript:c.POS REF > ALT`. \n",
    "Adding this informaiton to our data ordinarily provide a 'one to many' matching of variants to multiple affected transcripts. To avoid massively inflating the size of the output data, and since we are currently investigating variants instead of transcripts, I will instead concat HGVS consequences together into a single, commda-delimited 'hgvsc' column in the output. This is a nice balance: we don't massively inflate the size of our data, but the hgvsc information is available for reference in the output.\n",
    "\n",
    "\n",
    "One **caution** before we proceed: when querying VEP, we need to ensure we are using the correct reference genome (hg19/GRCh37 versus hg38/GRCh38) as many genomic positions differ from one build to another. Looking at the VCF header, we can see a reference to 'human_g1k_v37.fasta,' which is a GRCh37 reference genome build from the 1k Genomics project. With this in mind, all of our queries should use the server https://grch37.rest.ensembl.org instead of the more recent GRCh38 server.\n",
    "\n",
    "The first challenge is performing batch queries of the API on what is otherwise a relatively large file (with over 10k variants). Let's begin by writing a function that queries the API in batches, and testing the function with a small subset of our data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "80abff6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Querying VEP: batch 1 / 2\n",
      "Querying VEP: batch 2 / 2\n",
      "Query completed!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>regulatory_feature_consequences</th>\n",
       "      <th>colocated_variants</th>\n",
       "      <th>id</th>\n",
       "      <th>end</th>\n",
       "      <th>allele_string</th>\n",
       "      <th>start</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>apiq</th>\n",
       "      <th>seq_region_name</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>transcript_consequences</th>\n",
       "      <th>motif_feature_consequences</th>\n",
       "      <th>intergenic_consequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>[{'consequence_terms': ['regulatory_region_var...</td>\n",
       "      <td>[{'end': 108093328, 'id': 'rs9480798', 'minor_...</td>\n",
       "      <td>.</td>\n",
       "      <td>108093328</td>\n",
       "      <td>G/A</td>\n",
       "      <td>108093328</td>\n",
       "      <td>SNV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>6 108093328 . G A . . .</td>\n",
       "      <td>6</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[{'gene_symbol': 'SCML4', 'consequence_terms':...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'allele_string': 'C/G/T', 'strand': 1, 'seq_...</td>\n",
       "      <td>.</td>\n",
       "      <td>134760584</td>\n",
       "      <td>C/T</td>\n",
       "      <td>134760584</td>\n",
       "      <td>SNV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10 134760584 . C T . . .</td>\n",
       "      <td>10</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[{'variant_allele': 'T', 'gene_id': 'ENSG00000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'start': 77450964, 'var_synonyms': {'COSMIC'...</td>\n",
       "      <td>.</td>\n",
       "      <td>77450964</td>\n",
       "      <td>C/T</td>\n",
       "      <td>77450964</td>\n",
       "      <td>SNV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>15 77450964 . C T . . .</td>\n",
       "      <td>15</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[{'hgnc_id': 29431, 'sift_prediction': 'tolera...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>[{'regulatory_feature_id': 'ENSR00000309957', ...</td>\n",
       "      <td>[{'id': 'rs80219351', 'end': 54179796, 'start'...</td>\n",
       "      <td>.</td>\n",
       "      <td>54179796</td>\n",
       "      <td>G/A</td>\n",
       "      <td>54179796</td>\n",
       "      <td>SNV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>4 54179796 . G A . . .</td>\n",
       "      <td>4</td>\n",
       "      <td>splice_region_variant</td>\n",
       "      <td>[{'hgvsc': 'ENST00000388940.4:c.1135+7C&gt;T', 'i...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'frequencies': {'G': {'amr': 0.987, 'eas': 1...</td>\n",
       "      <td>.</td>\n",
       "      <td>18470980</td>\n",
       "      <td>A/G</td>\n",
       "      <td>18470980</td>\n",
       "      <td>SNV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>20 18470980 . A G . . .</td>\n",
       "      <td>20</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[{'transcript_id': 'ENST00000337227.4', 'gene_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   strand                    regulatory_feature_consequences  \\\n",
       "0       1  [{'consequence_terms': ['regulatory_region_var...   \n",
       "1       1                                                NaN   \n",
       "2       1                                                NaN   \n",
       "3       1  [{'regulatory_feature_id': 'ENSR00000309957', ...   \n",
       "4       1                                                NaN   \n",
       "\n",
       "                                  colocated_variants id        end  \\\n",
       "0  [{'end': 108093328, 'id': 'rs9480798', 'minor_...  .  108093328   \n",
       "1  [{'allele_string': 'C/G/T', 'strand': 1, 'seq_...  .  134760584   \n",
       "2  [{'start': 77450964, 'var_synonyms': {'COSMIC'...  .   77450964   \n",
       "3  [{'id': 'rs80219351', 'end': 54179796, 'start'...  .   54179796   \n",
       "4  [{'frequencies': {'G': {'amr': 0.987, 'eas': 1...  .   18470980   \n",
       "\n",
       "  allele_string      start variant_class assembly_name  \\\n",
       "0           G/A  108093328           SNV        GRCh37   \n",
       "1           C/T  134760584           SNV        GRCh37   \n",
       "2           C/T   77450964           SNV        GRCh37   \n",
       "3           G/A   54179796           SNV        GRCh37   \n",
       "4           A/G   18470980           SNV        GRCh37   \n",
       "\n",
       "                       apiq seq_region_name most_severe_consequence  \\\n",
       "0   6 108093328 . G A . . .               6          intron_variant   \n",
       "1  10 134760584 . C T . . .              10          intron_variant   \n",
       "2   15 77450964 . C T . . .              15        missense_variant   \n",
       "3    4 54179796 . G A . . .               4   splice_region_variant   \n",
       "4   20 18470980 . A G . . .              20          intron_variant   \n",
       "\n",
       "                             transcript_consequences  \\\n",
       "0  [{'gene_symbol': 'SCML4', 'consequence_terms':...   \n",
       "1  [{'variant_allele': 'T', 'gene_id': 'ENSG00000...   \n",
       "2  [{'hgnc_id': 29431, 'sift_prediction': 'tolera...   \n",
       "3  [{'hgvsc': 'ENST00000388940.4:c.1135+7C>T', 'i...   \n",
       "4  [{'transcript_id': 'ENST00000337227.4', 'gene_...   \n",
       "\n",
       "  motif_feature_consequences intergenic_consequences  \n",
       "0                        NaN                     NaN  \n",
       "1                        NaN                     NaN  \n",
       "2                        NaN                     NaN  \n",
       "3                        NaN                     NaN  \n",
       "4                        NaN                     NaN  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from math import ceil\n",
    "import json\n",
    "import requests\n",
    "\n",
    "# write a quick function to chunk our data into 200-variant sizes\n",
    "def chunker(seq, size):\n",
    "    # chunk data into n sizes\n",
    "    return (seq[pos:pos + size] for pos in range(0, len(seq), size))\n",
    "\n",
    "# write a quick wrapper function to query the API in batches\n",
    "def vepquery(df, size=200):\n",
    "    \n",
    "    # This function takes a VCF-style input and performs a region query \n",
    "    # to obtain overlapping biological regions (genes and trascripts)\n",
    "    \n",
    "    # cut our dataframe into chunks\n",
    "    nchunks = ceil(df.shape[0] / size)\n",
    "    n=1\n",
    "    decoded=pd.DataFrame()\n",
    "    for chunk in chunker(df, size):\n",
    "        \n",
    "        print('Querying VEP: batch', n, '/', nchunks )\n",
    "    \n",
    "        # format a query of our variants\n",
    "        query = chunk.apiq.to_list()\n",
    "        query = json.dumps(\n",
    "            {\"variants\":query})\n",
    "        \n",
    "        params = {'transcript_version':'true', 'variant_class':'true', 'hgvs':'true'}\n",
    "        \n",
    "        # query the VEP API\n",
    "        server = \"https://grch37.rest.ensembl.org\"\n",
    "        ext = \"/vep/homo_sapiens/region\"\n",
    "        headers={ \"Content-Type\" : \"application/json\", \"Accept\" : \"application/json\"}\n",
    "        r = requests.post(server+ext, headers=headers, data=query, params=params)\n",
    "        if not r.ok:\n",
    "          r.raise_for_status()\n",
    "          sys.exit()\n",
    "         \n",
    "        # decode output\n",
    "        if decoded.empty:\n",
    "            \n",
    "            # decode to pandas df\n",
    "            decoded = pd.DataFrame(r.json()).rename(columns={'input':'apiq'})\n",
    "            \n",
    "        else:\n",
    "            decoded = pd.concat([decoded,\n",
    "                                 pd.DataFrame(r.json()).rename(columns={'input':'apiq'})])\n",
    "            \n",
    "        n+=1\n",
    "        \n",
    "    return (decoded)\n",
    "\n",
    "# now let's take a small sample of our overall variants for demo purposes\n",
    "# (running on the whole dataset takes a long time)\n",
    "df = df.sample(n=250)\n",
    "\n",
    "# format an API query column in our data that VEP will understand as input\n",
    "df['apiq'] = df.apply(\n",
    "    lambda row: ' '.join([str(i) for i in [row.CHROM, ' ', row.POS, \" .\", row.REF, row.ALT, \".\", \".\", \".\"]]), axis=1)\n",
    "\n",
    "# run VEP query\n",
    "decoded = vepquery(df, 200)\n",
    "print('Query completed!')\n",
    "\n",
    "decoded.head(n=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ebf1a74",
   "metadata": {},
   "source": [
    "This VEP output returns lots of information that we are not (at this time) interested in using. So let's clean up the output and extract our columns of interest. *Side note: this is going to require some pandas method chaining which is not as readable, just know that this is some nuts and bolts formatting to unpacking the json returned from VEP.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "64cb6c17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# take our VEP results and separate out transcript consequences to their own rows\n",
    "# yes, this inflates the size of the data, but we will undo this in a few lines\n",
    "decoded = decoded.explode('transcript_consequences')\n",
    "\n",
    "\n",
    "# write some helper functions to assist with json unpacking\n",
    "def collapse_list(series):\n",
    "    # collapse lists in a pandas series to a single str\n",
    "    return ( ','.join(\n",
    "        (series.apply(pd.Series).stack().reset_index(drop=True).dropna().unique())))\n",
    "\n",
    "def checktype(series):\n",
    "    # check types in a pandas series, return dtype if there is only one dtype\n",
    "    types = series.apply(type).unique()\n",
    "    if types.size == 1:\n",
    "        return(types[0])\n",
    "    else:\n",
    "        return(object)\n",
    "\n",
    "def annot_extract(df, target_cell, newname='annot'):\n",
    "\n",
    "    # extract nested json annotations from a target cell\n",
    "    # return copy of dataframee with an added 'newname' col\n",
    "\n",
    "    # apply json_norm to cell\n",
    "    col = df.loc[df[target_cell].notna()][target_cell].apply(pd.json_normalize)\n",
    "\n",
    "    # use list comprehension + lambda fn to obtain json info in a single series\n",
    "    annot = pd.Series (\n",
    "        col.apply( lambda cell:\n",
    "        {i : (';'.join(cell[i].astype(str).unique()) ) \n",
    "         if checktype( cell[i])!=list else collapse_list(cell[i]) \n",
    "         for i in cell.columns} ) )\n",
    "\n",
    "    # add info to new column, propagate nan values\n",
    "    df[newname] = np.nan\n",
    "    df.loc[df[target_cell].notna(), newname] = annot\n",
    "\n",
    "    return(df)\n",
    "\n",
    "\n",
    "# extract transcript consequences\n",
    "decoded = annot_extract(decoded, 'transcript_consequences')\n",
    "\n",
    "# explode values to new column\n",
    "decoded['gene_symbol'] = decoded.annot.apply(pd.Series)['gene_symbol']\n",
    "decoded['gene_id'] = decoded.annot.apply(pd.Series)['gene_id']\n",
    "decoded['effect'] = decoded.annot.apply(pd.Series)['consequence_terms']\n",
    "decoded['hgvsc'] = decoded.annot.apply(pd.Series)['hgvsc']\n",
    "\n",
    "# subset to cols of interest\n",
    "decoded = decoded[['apiq', 'variant_class', 'gene_symbol', 'gene_id', 'effect', 'hgvsc']]\n",
    "\n",
    "# now that we have our annotations of interest, we can recompress them so we have one line per variant/gene\n",
    "# groupby variant info (apiq) and gene_id, then\n",
    "# compress hgvsc and effect annotations into a delimited string\n",
    "hgvsc = decoded.loc[decoded.hgvsc.notna()].groupby(['apiq','gene_id','variant_class', 'gene_symbol'])['hgvsc'].unique().apply(lambda x: ','.join((x))).reset_index()[['apiq','gene_id','hgvsc']]\n",
    "decoded = decoded.groupby(['apiq','gene_id','variant_class', 'gene_symbol'])['effect'].unique().apply(lambda x: ','.join((x))).reset_index()\n",
    "\n",
    "# merge hgvsc back into our slimmed dataframe\n",
    "decoded = decoded.merge(hgvsc, on=['apiq','gene_id'], how='left')\n",
    "\n",
    "# handle duplicates in effect and hgvsc cells (I believe caused by output from VEP, not concatenation process in pandas)\n",
    "decoded.loc[:, 'effect'] = decoded.effect.apply(lambda x: ','.join(np.unique(x.split(','))))\n",
    "decoded.loc[decoded.hgvsc.notna(), 'hgvsc'] = decoded.loc[decoded.hgvsc.notna()].hgvsc.apply(lambda x: ','.join(np.unique(x.split(','))))\n",
    "\n",
    "\n",
    "# fill nan\n",
    "decoded.effect.fillna('intergenic', inplace=True)\n",
    "decoded.fillna('None', inplace=True)\n",
    "\n",
    "# merge back to original df\n",
    "decoded.loc[:, 'apiq'] = decoded.apiq.str.replace(' ','')\n",
    "df.loc[:, 'apiq'] = df.apiq.str.replace(' ','')\n",
    "df = df.merge(decoded, on='apiq').drop(columns='apiq')\n",
    "\n",
    "# select cols of interest\n",
    "cols = ['CHROM','POS', 'REF', 'ALT', 'depth', 'var_reads', 'var_perc',\n",
    "        'variant_class', 'gene_symbol', 'gene_id', 'effect', 'hgvsc']\n",
    "df = df[cols]\n",
    "\n",
    "df.head(n=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40644894",
   "metadata": {},
   "source": [
    "With our intiial VEP query complete, we now have information on affected genes and transcripts, as well as our VCF variants in HGVS format. We are missing one item: an estimate of minor allele frequency (MAF). \n",
    "\n",
    "Unfortunately, we can't unpack this from our earlier VEP query, as no MAF information was returned. We can instead query allele frequency information from a separate VEP URL: `https://grch37.rest.ensembl.org/vep/homo_sapiens/hgvs`. While this is also going to run slowly, it's still one of the best options we have for retrieving MAF within the confines of an API (versus doing queries of a locally-downloaded MAF database such as dbSNP or GnomAD).\n",
    "\n",
    "Let's create (and implement) another quick wrapper function for MAF queries. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "89f0d68a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Querying MAF: batch 1 / 2\n",
      "Querying MAF: batch 2 / 2\n",
      "MAF query complete.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/54/1msgdf6x1qbgn696gqpw64600000gn/T/ipykernel_47369/532283771.py:53: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  freq = maf.set_index('apiq',drop=True).colocated_variants.apply(pd.Series).stack().apply(pd.Series).minor_allele_freq.reset_index()[['apiq', 'minor_allele_freq']]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>depth</th>\n",
       "      <th>var_reads</th>\n",
       "      <th>var_perc</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>effect</th>\n",
       "      <th>hgvsc</th>\n",
       "      <th>minor_allele_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>108093328</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>127</td>\n",
       "      <td>126</td>\n",
       "      <td>0.992126</td>\n",
       "      <td>SNV</td>\n",
       "      <td>SCML4</td>\n",
       "      <td>ENSG00000146285</td>\n",
       "      <td>NMD_transcript_variant,intron_variant,non_codi...</td>\n",
       "      <td>ENST00000369020.3:c.156+48C&gt;T,ENST00000369022....</td>\n",
       "      <td>0.0517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>134760584</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>332</td>\n",
       "      <td>173</td>\n",
       "      <td>0.521084</td>\n",
       "      <td>SNV</td>\n",
       "      <td>TTC40</td>\n",
       "      <td>ENSG00000171811</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>134760584</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>332</td>\n",
       "      <td>173</td>\n",
       "      <td>0.521084</td>\n",
       "      <td>SNV</td>\n",
       "      <td>LINC01166</td>\n",
       "      <td>ENSG00000232903</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant</td>\n",
       "      <td>ENST00000443633.1:n.105-599G&gt;A</td>\n",
       "      <td>0.4026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>77450964</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>194</td>\n",
       "      <td>194</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SNV</td>\n",
       "      <td>PEAK1</td>\n",
       "      <td>ENSG00000173517</td>\n",
       "      <td>missense_variant,non_coding_transcript_exon_va...</td>\n",
       "      <td>ENST00000312493.4:c.3212G&gt;A,ENST00000560626.2:...</td>\n",
       "      <td>0.1981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>54179796</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>272</td>\n",
       "      <td>130</td>\n",
       "      <td>0.477941</td>\n",
       "      <td>SNV</td>\n",
       "      <td>SCFD2</td>\n",
       "      <td>ENSG00000184178</td>\n",
       "      <td>intron_variant,splice_region_variant</td>\n",
       "      <td>ENST00000388940.4:c.1135+7C&gt;T,ENST00000401642....</td>\n",
       "      <td>0.0110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM        POS REF ALT depth var_reads  var_perc variant_class  \\\n",
       "0     6  108093328   G   A   127       126  0.992126           SNV   \n",
       "1    10  134760584   C   T   332       173  0.521084           SNV   \n",
       "2    10  134760584   C   T   332       173  0.521084           SNV   \n",
       "3    15   77450964   C   T   194       194  1.000000           SNV   \n",
       "4     4   54179796   G   A   272       130  0.477941           SNV   \n",
       "\n",
       "  gene_symbol          gene_id  \\\n",
       "0       SCML4  ENSG00000146285   \n",
       "1       TTC40  ENSG00000171811   \n",
       "2   LINC01166  ENSG00000232903   \n",
       "3       PEAK1  ENSG00000173517   \n",
       "4       SCFD2  ENSG00000184178   \n",
       "\n",
       "                                              effect  \\\n",
       "0  NMD_transcript_variant,intron_variant,non_codi...   \n",
       "1                              upstream_gene_variant   \n",
       "2       intron_variant,non_coding_transcript_variant   \n",
       "3  missense_variant,non_coding_transcript_exon_va...   \n",
       "4               intron_variant,splice_region_variant   \n",
       "\n",
       "                                               hgvsc  minor_allele_freq  \n",
       "0  ENST00000369020.3:c.156+48C>T,ENST00000369022....             0.0517  \n",
       "1                                               None             0.0000  \n",
       "2                     ENST00000443633.1:n.105-599G>A             0.4026  \n",
       "3  ENST00000312493.4:c.3212G>A,ENST00000560626.2:...             0.1981  \n",
       "4  ENST00000388940.4:c.1135+7C>T,ENST00000401642....             0.0110  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define a wrapper function (very similar to our previous function)\n",
    "def mafquery(df, size=200):\n",
    "    \n",
    "    nchunks = ceil(df.shape[0] / size)\n",
    "    n=1\n",
    "    decoded=pd.DataFrame()\n",
    "    \n",
    "    for chunk in chunker(df, size):\n",
    "        \n",
    "        print('Querying MAF: batch', n, '/', nchunks )\n",
    "    \n",
    "        # format a query of our variants\n",
    "        query = list(chunk.apiq.unique())\n",
    "        query = json.dumps(\n",
    "            {\"hgvs_notations\":query}\n",
    "            )\n",
    "        \n",
    "        # query the VEP API\n",
    "        server = \"https://grch37.rest.ensembl.org\"\n",
    "        ext = \"/vep/homo_sapiens/hgvs\"\n",
    "        headers={ \"Content-Type\" : \"application/json\", \"Accept\" : \"application/json\"}\n",
    "        r = requests.post(server+ext, headers=headers, data=query)\n",
    "        if not r.ok:\n",
    "          r.raise_for_status()\n",
    "          sys.exit()\n",
    "         \n",
    "        # decode output\n",
    "        if decoded.empty:\n",
    "            \n",
    "            # decode to pandas df\n",
    "            decoded = pd.DataFrame(r.json()).rename(columns={'input':'apiq'})\n",
    "            \n",
    "        else:\n",
    "            decoded = pd.concat([decoded,\n",
    "                                 pd.DataFrame(r.json()).rename(columns={'input':'apiq'})])\n",
    "            \n",
    "        n+=1\n",
    "        \n",
    "    return (decoded)\n",
    "\n",
    "\n",
    "# reformat our API query to use the HGVS consequences\n",
    "df['apiq'] = df.hgvsc.str.split(',').str[0]\n",
    "\n",
    "# query MAF information using our wrapper function\n",
    "maf = mafquery(df[['CHROM','apiq']].drop_duplicates())\n",
    "print('Formatting results.')\n",
    "\n",
    "# drop entries within colocated_variants that do not contain frequencies info\n",
    "maf.loc[maf.colocated_variants.notna(), 'colocated_variants'] = maf.colocated_variants.dropna().apply(\n",
    "    lambda x: [i for i in x if 'frequencies' in i.keys()])\n",
    "\n",
    "# stack all frequency data\n",
    "freq = maf.set_index('apiq',drop=True).colocated_variants.apply(pd.Series).stack().apply(pd.Series).minor_allele_freq.reset_index()[['apiq', 'minor_allele_freq']]\n",
    "freq = freq.dropna().drop_duplicates()\n",
    "\n",
    "# merge MAF into df, fill missing entries with MAF of zero\n",
    "maf = maf.merge(freq, how='left', on='apiq')\n",
    "maf = maf[['apiq', 'minor_allele_freq']].drop_duplicates()\n",
    "\n",
    "# merge MAF info back to our main df, fill missing values\n",
    "df = df.merge(maf, how='left', on='apiq')\n",
    "df.loc[:, 'minor_allele_freq'] = df.minor_allele_freq.fillna(0)\n",
    "\n",
    "# subset to output cols\n",
    "df = df[ cols+['minor_allele_freq'] ]\n",
    "\n",
    "print('MAF query complete.')\n",
    "\n",
    "df.head(n=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0449ad72",
   "metadata": {},
   "source": [
    "With that, our mini-pipeline is completed! Our output contains the following:\n",
    "+ **VCF input** (CHROM, POS, REF, ALT).\n",
    "+ **depth:** read depth at site.\n",
    "+ **var_reads:** variant supporting reads at site.\n",
    "+ **var_perc:** percent of reads at site supporting variant.\n",
    "+ **variant_class:** type of variant (ie SNV, indel)\n",
    "+ **gene_id:** Ensembl gene identifier.\n",
    "+ **gene symbol:** more familiar gene symbol/abbreviation.\n",
    "+ **effect:** variant effect on transscript (ie intronic, missense, etc).\n",
    "+ **hgvsc:** comma-delimited list of all hgvsc variants at site (affected transcript+molecular consequences.\n",
    "+ **minor_allele_freq:** minor allele frequency from VEP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca951309",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
